src/ServiceControl/Infrastructure/RavenDB/Expiration/ExpiredDocumentsCleaner.cs

```diff
@@ -111,71 +92,66 @@ void Delete(PeriodicExecutor executor)
                             }
 
                             var id = doc.Value<string>("__document_id");
-                            if (!string.IsNullOrEmpty(id))
+                            if (string.IsNullOrEmpty(id))
                             {
-                                items.Add(new DeleteCommandData
-                                {
-                                    Key = id
-                                });
+                                return;
+                            }
+                            items.Add(new DeleteCommandData
+                            {
+                                Key = id
+                            });
 
-                                var bodyNotStored = doc.SelectToken("MessageMetadata.BodyNotStored", errorWhenNoMatch: false);
-                                if (bodyNotStored == null || bodyNotStored.Value<bool>() == false)
-                                {
-                                    var msgId = doc.SelectToken("MessageMetadata.MessageId", errorWhenNoMatch: false);
-                                    if (msgId != null)
-                                    {
-                                        var attachmentId = "messagebodies/" + msgId.Value<string>();
-                                        attachments.Add(attachmentId);
-                                    }
-                                }
+                            string bodyId;
+                            if (TryGetBodyId(doc, out bodyId))
+                            {
+                                attachments.Add(bodyId);
                             }
                         });
-                    }
-                    catch (OperationCanceledException)
-                    {
-                        //Ignore
-                    }
-
-                    logger.Debug("Batching deletion of {0} documents.",items.Count);
-
-                    docsToExpire += items.Count;
-                    var results = Database.Batch(items.ToArray());
-                    Database.TransactionalStorage.Batch(accessor =>
-                    {
-                        foreach (var attach in attachments)
-                        {
-                            accessor.Attachments.DeleteAttachment(attach, null);
-                        }
-                    });
-                    deletionCount = results.Count(x => x.Deleted == true);
-                    items.Clear();
                 }
-
-                if (docsToExpire == 0)
+                catch (OperationCanceledException)
                 {
-                    logger.Debug("No expired documents found");
+                    //Ignore
                 }
-                else
+
+                logger.Debug("Batching deletion of {0} documents.", items.Count);
+
+                docsToExpire += items.Count;
```

Comment:
docsToExpire is scoped outside the database.Query so it increments on each run.

Am i missing something or is your suggestion invalid?